ProcessLib/BoundaryCondition/ConstraintDirichletBoundaryCondition.cpp

```diff
@@ -0,0 +1,324 @@
+        "constraints");
+    for (auto const* boundary_element : _bc_mesh.getElements())
+    {
+        _flux_values[boundary_element->getID()] =
+            _local_assemblers[boundary_element->getID()]->integrate(
+                x, t, _bulk_mesh,
+                [this](std::size_t const element_id,
+                       MathLib::Point3d const& pnt, double const t,
+                       GlobalVector const& x) {
+                    return _getFlux(element_id, pnt, t, x);
+                });
+    }
+}
+
+void ConstraintDirichletBoundaryCondition::getEssentialBCValues(
+    const double t, NumLib::IndexValueVector<GlobalIndexType>& bc_values) const
+{
+    SpatialPosition pos;
+
+    bc_values.ids.clear();
+    bc_values.values.clear();
+
+    std::vector<std::pair<GlobalIndexType, double>> tmp_bc_values;
+
+    auto isFlux = [&](const std::size_t element_id) {
+        return _lower ? _flux_values[element_id] < _constraint_threshold
+                      : _flux_values[element_id] > _constraint_threshold;
+    };
+
+    for (auto const* boundary_element : _bc_mesh.getElements())
+    {
+        // check if the boundary element is active
+        if (isFlux(boundary_element->getID()))
+        {
+            continue;
+        }
+
+        // loop over the boundary element nodes and set the dirichlet value
+        unsigned const number_nodes = boundary_element->getNumberOfNodes();
+        for (unsigned i = 0; i < number_nodes; ++i)
+        {
+            auto const id = boundary_element->getNode(i)->getID();
+            pos.setNodeID(id);
+
+            MeshLib::Location l(_bulk_mesh.getID(), MeshLib::MeshItemType::Node,
+                                id);
+            // TODO: that might be slow, but only done once
+            const auto g_idx = _dof_table_boundary->getGlobalIndex(
+                l, _variable_id, _component_id);
+            if (g_idx == NumLib::MeshComponentMap::nop)
+                continue;
+            // For the DDC approach (e.g. with PETSc option), the negative
+            // index of g_idx means that the entry by that index is a ghost one,
+            // which should be dropped. Especially for PETSc routines
+            // MatZeroRows and MatZeroRowsColumns, which are called to apply the
+            // Dirichlet BC, the negative index is not accepted like other
+            // matrix or vector PETSc routines. Therefore, the following
+            // if-condition is applied.
+            if (g_idx >= 0)
+            {
+                tmp_bc_values.emplace_back(g_idx, _parameter(t,pos).front());
+            }
+        }
+    }
+
+    if (tmp_bc_values.empty())
+    {
+        DBUG("The domain on which the boundary is defined is empty");
+        return;
+    }
+
+    // make the pairs (node id, value) unique
+    // first: sort the (node id, value) pairs according to the node id
+    std::sort(tmp_bc_values.begin(), tmp_bc_values.end(),
```

Comment:
Sorry, but I still don't see where the tmp_bc_values are "made unique". There is a copy, but not a std::unique followed by erase....

Reply:
I changed the comment.